From: Hannes Reinecke <hare@suse.de>
Date: Thu, 29 Mar 2018 10:38:37 +0200
Subject: [PATCH] nvme/rdma: do no start error recovery twice
References: bsc#1084967
Patch-Mainline: never, SLE12-SP3 specific

If error recovery is already scheduled we should just ignore
all other errors; they'll be cleared up when resetting the
controller.

Reported-by: Steve Schremmer <sschremm@netapp.com>
Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/nvme/host/rdma.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/nvme/host/rdma.c b/drivers/nvme/host/rdma.c
index 608b0b59dde8..e13509970e24 100644
--- a/drivers/nvme/host/rdma.c
+++ b/drivers/nvme/host/rdma.c
@@ -849,6 +849,10 @@ static void nvme_rdma_error_recovery_work(struct work_struct *work)
 
 static void nvme_rdma_error_recovery(struct nvme_rdma_ctrl *ctrl)
 {
+	/* Error recovery should trigger on the first error */
+	if (ctrl->ctrl.state != NVME_CTRL_LIVE)
+		return;
+
 	if (!nvme_change_ctrl_state(&ctrl->ctrl, NVME_CTRL_RESETTING))
 		return;
 
-- 
2.12.3

